src/decode/stream.rs

```diff
@@ -0,0 +1,476 @@
+{
+    fn new(decoder: DecoderState<W, LZCircularBuffer<W>>, range: u32, code: u32) -> Self {
+        Self {
+            decoder,
+            range,
+            code,
+        }
+    }
+}
+
+impl<W> std::fmt::Debug for RunState<W>
+where
+    W: Write,
+{
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::result::Result<(), std::fmt::Error> {
+        fmt.debug_struct("RunState")
+            .field("range", &self.range)
+            .field("code", &self.code)
+            .finish()
+    }
+}
+
+/// Lzma decompressor that can process multiple chunks of data using the
+/// `std::io::Write` interface.
+pub struct Stream<W>
+where
+    W: Write,
+{
+    /// Temporary buffer to hold data while the header is being read.
+    tmp: [u8; MAX_TMP_LEN],
+    /// How many bytes of the temp buffer are in use.
+    tmp_len: usize,
+    /// Whether the stream is initialized and ready to process data.
+    /// An `Option` is used to avoid interior mutability when updating the state.
+    state: Option<State<W>>,
+    /// Options given when a stream is created.
+    options: Options,
+}
+
+impl<W> Stream<W>
+where
+    W: Write,
+{
+    /// Initialize the stream. This will consume the `output` which is the sink
+    /// implementing `std::io::Write` that will receive decompressed bytes.
+    pub fn new(output: W) -> Self {
+        Self::new_with_options(Options::default(), output)
+    }
+
+    /// Initialize the stream with the given `options`. This will consume the
+    /// `output` which is the sink implementing `std::io::Write` that will
+    /// receive decompressed bytes.
+    pub fn new_with_options(options: Options, output: W) -> Self {
+        Self {
+            tmp: [0; MAX_TMP_LEN],
+            tmp_len: 0,
+            state: Some(State::Init(output)),
+            options,
+        }
+    }
+
+    /// Get a reference to the output sink
+    pub fn get_ref(&self) -> Option<&W> {
+        match &self.state {
+            Some(State::Init(output)) => Some(&output),
+            Some(State::Run(state)) => Some(state.decoder.output.get_ref()),
+            None => None,
+        }
+    }
+
+    /// Get a mutable reference to the output sink
+    pub fn get_mut(&mut self) -> Option<&mut W> {
+        match &mut self.state {
+            Some(State::Init(output)) => Some(output),
+            Some(State::Run(state)) => Some(state.decoder.output.get_mut()),
+            None => None,
+        }
+    }
+
+    /// Consumes the stream and returns the output sink. This also makes sure
+    /// we have properly reached the end of the stream.
+    pub fn finish(mut self) -> Result<W, std::io::Error> {
+        if let Some(state) = self.state.take() {
+            match state {
+                State::Init(output) => {
+                    if self.tmp_len > 0 {
+                        Err(std::io::Error::new(
+                            std::io::ErrorKind::Other,
+                            "failed to read header",
+                        ))
```

Comment:
This should be an `LZMAError`.

Reply:
I had to implement `Display` for `lzma_rs::error::Error`. I put this in a separate PR #53 .